### PR TITLE
CDF-18450: add tracing support

### PR DESCRIPTION
**Pull request by dmivankov**
Adding tracing propagation support.

1. Introduce `type TracedIO = Kleisli[IO, Span[IO], _]` and some helper methods for lifting to it from IO, also helpers for passing parent span from Config or across partitioning calls
2. Introduce some operation spans, more detailed spans to be added later
3. Add support for passing parent tracing context via `TRACING_PARAMETER_PREFIX: String = "com.cognite.tracing.parameter."` prefix within parameters
4. Add env vars `OTEL_COLLECTOR_URL`, and optional `OTEL_SERVICE_NAME`, `OTEL_SERVICE_VERSION` that trigger span reporting
5. Bulk of lines changed is just doing IO->TracedIO and adding new span for places with unsafeRunSync

Should be safe to use without tracing as before. To use tracing
- start passing Kernel/parent context parameters with `TRACING_PARAMETER_PREFIX`
- add `OTEL_COLLECTOR_URL` env var to enable trace info reporting

[CDF-18450]

[CDF-18450]: https://cognitedata.atlassian.net/browse/CDF-18450?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/scala/cognite/spark/v1/FlexibleDataModelRelationFactory.scala [hunk 1]
```diff
@@ -81,74 +81,79 @@ object FlexibleDataModelRelationFactory {
           case _ => false
         }
       }
-      .flatMap {
-        case Some(vc: ViewDefinition) => IO.pure(Some(vc))
+      .flatMap[Option[DataModelViewReference], Span[IO]] {
+        case Some(vc: ViewDefinition) => TracedIO.pure(Some(vc: DataModelViewReference))
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/FlexibleDataModelRelationFactory.scala:85]:
```suggestion
        case Some(vc: ViewDefinition) => TracedIO.pure(Some(vc))
```

File: src/main/scala/cognite/spark/v1/FlexibleDataModelRelationFactory.scala [hunk 2]
```diff
@@ -81,74 +81,79 @@ object FlexibleDataModelRelationFactory {
...
         case Some(vr: ViewReference) => fetchViewWithAllProps(client, vr).map(_.headOption)
-        case _ => IO.pure(None)
+        case _ => TracedIO.pure(None)
       }
       .flatMap {
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/FlexibleDataModelRelationFactory.scala:89]:
```suggestion
      .flatMap[Option[DataModelViewReference], Span[IO]] {
```

File: src/main/scala/cognite/spark/v1/package.scala [hunk 3]
```diff
@@ -1,11 +1,36 @@
...
+    def delay[A](a: A): TracedIO[A] = Kleisli.liftF(IO.delay(a))
+
+    def childPure[B](parent: Span[IO], name: String)(f: Span[IO] => B): IO[B] =
+      child(parent, name)(Kleisli { span =>
+        IO.delay(f(span))
+      })
+
+    def child[B](parent: Span[IO], name: String)(f: TracedIO[B]): IO[B] =
+      parent.span(name).use(f.run)
+  }
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/package.scala:32]: new helpers here

File: src/main/scala/cognite/spark/v1/StreamIterator.scala [hunk 4]
```diff
@@ -61,19 +75,20 @@ object StreamIterator {
...
   def enqueueStreamResults[A](
-      stream: Stream[IO, A],
+      stream: Stream[TracedIO, A],
       queue: EitherQueue[A],
       queueBufferSize: Int,
-      processChunk: Option[Chunk[A] => Chunk[A]]): Stream[IO, Unit] =
+      processChunk: Option[Chunk[A] => Chunk[A]]): Stream[TracedIO, Unit] =
     stream.chunks.parEvalMapUnordered(queueBufferSize) { chunk =>
-      IO.blocking {
+      Kleisli.liftF(IO.blocking {
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/StreamIterator.scala:88]:
```suggestion
      TracedIO.liftIOIO.blocking {
```

File: src/main/scala/cognite/spark/v1/CdfSparkAuth.scala [hunk 5]
```diff
@@ -1,39 +1,46 @@
...
 
 object CdfSparkAuth {
   final case class Static(auth: Auth) extends CdfSparkAuth {
     override def provider(
-        implicit clock: Clock[IO],
-        sttpBackend: SttpBackend[IO, Any]): IO[AuthProvider[IO]] =
-      IO(AuthProvider(auth))
+        implicit clock: Clock[TracedIO],
+        sttpBackend: SttpBackend[TracedIO, Any]): TracedIO[AuthProvider[TracedIO]] =
+      Kleisli.liftF(IO(AuthProvider[TracedIO](auth)))
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/CdfSparkAuth.scala:19]:
```suggestion
      TracedIO.liftIO(IO(AuthProvider[TracedIO](auth)))
```

File: src/main/scala/cognite/spark/v1/CdfSparkAuth.scala [hunk 6]
```diff
@@ -1,39 +1,46 @@
 package cognite.spark.v1
 
+import cats.data.Kleisli
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/CdfSparkAuth.scala:3]:
```suggestion
```

File: src/main/scala/cognite/spark/v1/SparkF.scala [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+                .use(span => op.run(span))
+                .unsafeRunSync()))
+      }
+    })
+  }
+}
+
+object SparkF {
+  def apply(df: DataFrame): DataFrameF = DataFrameF(df)
+}
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/SparkF.scala:27]: helper for foreachPartition to support tracing jump into partitions with less unsafeRunSync-s etc on caller side

**Review by BenChand — APPROVED** [review]

**Pull request merged**